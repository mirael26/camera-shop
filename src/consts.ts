export const AppUrl = {
  Main: '/',
  Catalog: '/catalog',
  Page: '/page_',
  Product: '/product',
  Cart: '/cart',
  NotFound: '/404',
  UnknownError: '/unknown_error',
  ServerUnavailable: '/server_unavailable',
} as const;

export const ApiUrl = {
  Promo: '/promo',
  Products: '/cameras',
  Reviews: '/reviews',
  Similar: '/similar',
  Promocode: '/coupons',
  Order: '/orders',
} as const;

export const Modal = {
  AddToCart: 'addToCartModal',
  DeleteFromCart: 'deleteFromCartModal',
  Review: 'reviewModal',
  OrderSuccess: 'orderSuccessModal',
} as const;

export const Tab = {
  Features: 'features',
  Description: 'description',
} as const;

export const Month = {
  0: 'января',
  1: 'февраля',
  2: 'марта',
  3: 'апреля',
  4: 'мая',
  5: 'июня',
  6: 'июля',
  7: 'августа',
  8: 'сентября',
  9: 'октября',
  10: 'ноября',
  11: 'декабря',
} as const;

export const Param = {
  Sort: 'sort',
  Order: 'order',
  Page: 'page',
  Tab: 'tab',
  PriceMin: 'price_min',
  PriceMax: 'price_max',
  Category: 'category',
  Level: 'level',
  Type: 'type',
} as const;

export const SortType = {
  Price: 'price',
  Rating: 'rating',
} as const;

export const SortOrder = {
  Asc: 'asc',
  Desc: 'desc',
} as const;

export const Category = {
  Camera: 'Фотоаппарат',
  VideoCamera: 'Видеокамера',
} as const;

export const Level = {
  Beginner: 'Нулевой',
  Amateur: 'Любительский',
  Professional: 'Профессиональный',
} as const;

export const CameraType = {
  Digital: 'Цифровая',
  Film: 'Плёночная',
  Instant: 'Моментальная',
  Collectible: 'Коллекционная',
} as const;
