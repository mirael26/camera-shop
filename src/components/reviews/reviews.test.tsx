import { fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { AppUrl } from '../../consts';
import { renderTestApp } from '../../test/helpers/render-test-app';
import { renderWithReduxAndRouter } from '../../test/helpers/render-with-redux-and-router';
import { bigReviewsMock, reviewsMock } from '../../test/mocks';
import Reviews from './reviews';

describe('Reviews', () => {
  test('Render correctly', () => {
    const reviews = renderWithReduxAndRouter(<Reviews/>, {
      initialState: {
        data: { reviews: reviewsMock }
      }
    });
    expect(reviews).toMatchSnapshot();
  });

  test('Render correct reviews count', () => {
    renderWithReduxAndRouter(<Reviews/>, {
      initialState: {
        data: { reviews: reviewsMock }
      }
    });
    const reviewCards = screen.getAllByTestId('review-card');
    expect(reviewCards).toHaveLength(2);
  });

  test('Display message if no reviews', () => {
    renderWithReduxAndRouter(<Reviews/>, {
      initialState: {
        data: { reviews: [] }
      }
    });
    const reviewCards = screen.queryAllByTestId('review-card');
    expect(reviewCards).toHaveLength(0);
    expect(screen.getByText<HTMLParagraphElement>(/Отзывов пока нет/i)).toBeInTheDocument();
  });

  test('Display message if no reviews', () => {
    renderWithReduxAndRouter(<Reviews/>, {
      initialState: {
        data: { reviews: [] }
      }
    });
    const reviewCards = screen.queryAllByTestId('review-card');
    expect(reviewCards).toHaveLength(0);
    expect(screen.getByText<HTMLParagraphElement>(/Отзывов пока нет/i)).toBeInTheDocument();
  });

  test('Show more reviews on click at showmore-button', () => {
    renderWithReduxAndRouter(<Reviews/>, {
      initialState: {
        data: { reviews: bigReviewsMock }
      }
    });
    let reviewCards = screen.getAllByTestId('review-card');
    expect(reviewCards).toHaveLength(3);
    fireEvent.click(screen.getByText<HTMLButtonElement>(/Показать больше отзывов/i));
    reviewCards = screen.getAllByTestId('review-card');
    expect(reviewCards).toHaveLength(5);
  });

  test('Open review-modal when click on new-review button', () => {
    renderTestApp(null, {
      route: `${AppUrl.Catalog}${AppUrl.Product}/1`,
      initialState: {
        data: { reviews: [] }
      }
    });
    const newReview = screen.getByText<HTMLButtonElement>(/Оставить свой отзыв/i);
    fireEvent.click(newReview);

    expect(screen.getByTestId('review-modal')).toBeInTheDocument();
  });
});
