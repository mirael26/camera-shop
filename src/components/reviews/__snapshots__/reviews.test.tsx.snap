// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews Render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page-content__section"
      >
        <section
          class="review-block"
          data-testid="reviews"
        >
          <div
            class="container"
          >
            <div
              class="page-content__headed"
            >
              <h2
                class="title title--h3"
              >
                Отзывы
              </h2>
              <button
                class="btn"
                type="button"
              >
                Оставить свой отзыв
              </button>
            </div>
            <ul
              class="review-block__list"
            >
              <li
                class="review-card"
                data-testid="review-card"
              >
                <div
                  class="review-card__head"
                >
                  <p
                    class="title title--h4"
                  >
                    Александр
                  </p>
                  <time
                    class="review-card__data"
                    data-testid="review-date"
                    datetime="2022-10-17"
                  >
                    17
                     
                    октября
                  </time>
                </div>
                <div
                  class="rate review-card__rate"
                >
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <p
                    class="visually-hidden"
                  >
                    Оценка: 
                    5
                  </p>
                </div>
                <ul
                  class="review-card__list"
                >
                  <li
                    class="item-list"
                  >
                    <span
                      class="item-list__title"
                    >
                      Достоинства:
                    </span>
                    <p
                      class="item-list__text"
                      data-testid="review-advantage"
                    >
                      Покупали недавно. Пока нареканий нет.
                    </p>
                  </li>
                  <li
                    class="item-list"
                  >
                    <span
                      class="item-list__title"
                    >
                      Недостатки:
                    </span>
                    <p
                      class="item-list__text"
                      data-testid="review-disadvantage"
                    >
                      Нет.
                    </p>
                  </li>
                  <li
                    class="item-list"
                  >
                    <span
                      class="item-list__title"
                    >
                      Комментарий:
                    </span>
                    <p
                      class="item-list__text"
                      data-testid="review-review"
                    >
                      Приобрела камеру ориентируясь на отзывы, но выявила множество несоответствий. Цена не оправдала ожидания
                    </p>
                  </li>
                </ul>
              </li>
              <li
                class="review-card"
                data-testid="review-card"
              >
                <div
                  class="review-card__head"
                >
                  <p
                    class="title title--h4"
                  >
                    Никита
                  </p>
                  <time
                    class="review-card__data"
                    data-testid="review-date"
                    datetime="2022-09-12"
                  >
                    12
                     
                    сентября
                  </time>
                </div>
                <div
                  class="rate review-card__rate"
                >
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      data-testid="full-star-icon"
                      xlink:href="#icon-full-star"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    height="16"
                    width="17"
                  >
                    <use
                      xlink:href="#icon-star"
                    />
                  </svg>
                  <p
                    class="visually-hidden"
                  >
                    Оценка: 
                    4
                  </p>
                </div>
                <ul
                  class="review-card__list"
                >
                  <li
                    class="item-list"
                  >
                    <span
                      class="item-list__title"
                    >
                      Достоинства:
                    </span>
                    <p
                      class="item-list__text"
                      data-testid="review-advantage"
                    >
                      Недорогая, за такую цену отличный вариант.
                    </p>
                  </li>
                  <li
                    class="item-list"
                  >
                    <span
                      class="item-list__title"
                    >
                      Недостатки:
                    </span>
                    <p
                      class="item-list__text"
                      data-testid="review-disadvantage"
                    >
                      Не рекомендую!
                    </p>
                  </li>
                  <li
                    class="item-list"
                  >
                    <span
                      class="item-list__title"
                    >
                      Комментарий:
                    </span>
                    <p
                      class="item-list__text"
                      data-testid="review-review"
                    >
                      Камера для начинающих, самое то. Но урезанный комплект и многое придётся докупать
                    </p>
                  </li>
                </ul>
              </li>
            </ul>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page-content__section"
    >
      <section
        class="review-block"
        data-testid="reviews"
      >
        <div
          class="container"
        >
          <div
            class="page-content__headed"
          >
            <h2
              class="title title--h3"
            >
              Отзывы
            </h2>
            <button
              class="btn"
              type="button"
            >
              Оставить свой отзыв
            </button>
          </div>
          <ul
            class="review-block__list"
          >
            <li
              class="review-card"
              data-testid="review-card"
            >
              <div
                class="review-card__head"
              >
                <p
                  class="title title--h4"
                >
                  Александр
                </p>
                <time
                  class="review-card__data"
                  data-testid="review-date"
                  datetime="2022-10-17"
                >
                  17
                   
                  октября
                </time>
              </div>
              <div
                class="rate review-card__rate"
              >
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <p
                  class="visually-hidden"
                >
                  Оценка: 
                  5
                </p>
              </div>
              <ul
                class="review-card__list"
              >
                <li
                  class="item-list"
                >
                  <span
                    class="item-list__title"
                  >
                    Достоинства:
                  </span>
                  <p
                    class="item-list__text"
                    data-testid="review-advantage"
                  >
                    Покупали недавно. Пока нареканий нет.
                  </p>
                </li>
                <li
                  class="item-list"
                >
                  <span
                    class="item-list__title"
                  >
                    Недостатки:
                  </span>
                  <p
                    class="item-list__text"
                    data-testid="review-disadvantage"
                  >
                    Нет.
                  </p>
                </li>
                <li
                  class="item-list"
                >
                  <span
                    class="item-list__title"
                  >
                    Комментарий:
                  </span>
                  <p
                    class="item-list__text"
                    data-testid="review-review"
                  >
                    Приобрела камеру ориентируясь на отзывы, но выявила множество несоответствий. Цена не оправдала ожидания
                  </p>
                </li>
              </ul>
            </li>
            <li
              class="review-card"
              data-testid="review-card"
            >
              <div
                class="review-card__head"
              >
                <p
                  class="title title--h4"
                >
                  Никита
                </p>
                <time
                  class="review-card__data"
                  data-testid="review-date"
                  datetime="2022-09-12"
                >
                  12
                   
                  сентября
                </time>
              </div>
              <div
                class="rate review-card__rate"
              >
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    data-testid="full-star-icon"
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    xlink:href="#icon-star"
                  />
                </svg>
                <p
                  class="visually-hidden"
                >
                  Оценка: 
                  4
                </p>
              </div>
              <ul
                class="review-card__list"
              >
                <li
                  class="item-list"
                >
                  <span
                    class="item-list__title"
                  >
                    Достоинства:
                  </span>
                  <p
                    class="item-list__text"
                    data-testid="review-advantage"
                  >
                    Недорогая, за такую цену отличный вариант.
                  </p>
                </li>
                <li
                  class="item-list"
                >
                  <span
                    class="item-list__title"
                  >
                    Недостатки:
                  </span>
                  <p
                    class="item-list__text"
                    data-testid="review-disadvantage"
                  >
                    Не рекомендую!
                  </p>
                </li>
                <li
                  class="item-list"
                >
                  <span
                    class="item-list__title"
                  >
                    Комментарий:
                  </span>
                  <p
                    class="item-list__text"
                    data-testid="review-review"
                  >
                    Камера для начинающих, самое то. Но урезанный комплект и многое придётся докупать
                  </p>
                </li>
              </ul>
            </li>
          </ul>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
